### DISCUSSION TITLE
txt file to a schem file

### DISCUSSION THREAD

**Discussion by Spursf** [maruohon/litematica#692] [Q&A]
Is it possible to make a schematic file using a txt file filled with info about block postions. I've got the example converted to a .schem but it fails to load.

For example:
1 0 0 oak_plank
2 0 0 oak_plank
3 0 0 oak_plank

**Comment by maruohon** [comment]
There is currently no converter from any kind of text format.

Can you explain why you would need this? Is it not easier to build stuff in-game? Or if you want to externally generate schematics, then there is no point doing it to a text format, you can just directly create the actual schematic file in NBT format. There are a number of libraries to create Litematica schematics, at least for Python and JavaScript if I remember right.

**Reply by Spursf** [reply]
Yes, the idea is to externally generate schematics. I'll have to look into some of these libaries and try to make something work. My program dumps block locations to a txt file hence the question.
Thanks.

**Reply by maruohon** [reply]
Check [litemapy](https://pypi.org/project/litemapy/) if you use Python, or this: https://github.com/mms0316/mms0316.github.io/blob/master/js/schematics.js if you use JavaScript.
Then there is also [this message](https://discord.com/channels/169369095538606080/566649314001158165/1097564948999184394) on my Discord with some alternative(?) Python implementation.

**Reply by Spursf** [reply]
Thank you I'll check them out.

**Reply by Spursf** [reply]
I'm baffled by the sponge schematic format. I'm aware its not fully supported, but does asnyone have any good exmaples the documentation is quite unclear (linked under). From my understanding it should be editable outside of the game?

https://github.com/SpongePowered/Schematic-Specification/blob/master/versions/schematic-3.md

I'll continue to play with litemapy but so far I've had no luck. (dont use python so its a pain)

**Reply by maruohon** [reply]
Wait so are you trying to write stuff to a Sponge schematic file and not a Litematica schematic file? In that case those Litematica schematic libraries probably do you no good. I have no idea if there are libraries for handling Sponge schematics, but I would suspect there must be (-> Google).

What language are you actually working in? You could also look at the Litematica schematic code (in the `liteloader_1.12.2` branch), basically some of the code inside the `schematic` and `schematic.container` packages. (Mostly for the serialization and de-serialization part, the schematic storage and internals are kinda designed for Litematica's needs and probably don't make sense as-is for any external/simpler programs.)

I think your best bet for an external schematic editor supporting the Sponge format is the Amulet program, which is the rewrite/continuation of the old MCEdit program for 1.13+ (also supporting older versions and Bedrock Edition and MCPE worlds AFAIK.) I don't actually know if it supports the Sponge format, but I assume it must, if it has any schematic functionality at this point.

**Comment by mms0316** [comment]
I'd advise to use Sponge v2 format (https://github.com/SpongePowered/Schematic-Specification/blob/master/versions/schematic-2.md) instead of v3, as it's supported by Litematica, WorldEdit and some other programs.

My library does handle Sponge v2 and (only) writes to a .schem file (Sponge v2 format) - you could use this page's source code to base yourself on:
https://mms0316.github.io/fakeschem.html

**Reply by mms0316** [reply]
Ah, and this page renders .schem, but in a very quick way, using map colors:
https://mms0316.github.io/preview.html

**Reply by Spursf** [reply]
Thanks

**Reply by Spursf** [reply]
@mms0316 You're tools have been helpful thus far. Is is possible to output block locations states and blocks into a txt file? The 3D viewer already knows the location, state, and type of the blocks in the file. If not I've made a request on the git repo. I'll look into it more and try to implement it myself If I get a chance.

**Reply by mms0316** [reply]
It's implemented now, so this issue can be closed.
